src/python/pants/option/options.py

```diff
@@ -364,17 +386,85 @@ def for_scope(
         values_builder = OptionValueContainerBuilder()
         flags_in_scope = self._scope_to_flags.get(scope, [])
         parse_args_request = self._make_parse_args_request(flags_in_scope, values_builder)
-        values = self.get_parser(scope).parse_args(
+        legacy_values = self.get_parser(scope).parse_args(
             parse_args_request, log_warnings=log_parser_warnings
         )
 
+        if self._native_options_validation == NativeOptionsValidation.IGNORE:
+            native_values = None
+        else:
+            native_values = self.get_parser(scope).parse_args_native(self._native_parser)
+
         # Check for any deprecation conditions, which are evaluated using `self._flag_matchers`.
         if check_deprecations:
-            values_builder = values.to_builder()
+            values_builder = legacy_values.to_builder()
             self._check_and_apply_deprecations(scope, values_builder)
-            values = values_builder.build()
+            legacy_values = values_builder.build()
+
+            if native_values:
+                native_values_builder = native_values.to_builder()
+                self._check_and_apply_deprecations(scope, native_values_builder)
+                native_values = native_values_builder.build()
+
+        def listify_tuples(x):
+            # Sequence values from the legacy parser can be tuple or list, but those coming from
+            # the native parser are always list. We convert to list here for comparison purposes.
+            if isinstance(x, (tuple, list)):
+                return [listify_tuples(y) for y in x]
+            elif isinstance(x, dict):
+                return {k: listify_tuples(v) for k, v in x.items()}
+            else:
+                return x
+
+        if native_values:
+            msgs = []
+
+            def legacy_val_info(k):
+                if k in legacy_values:
+                    val = legacy_values[k]
+                    descr = (
+                        f"{val} of type {type(val)}, from source {legacy_values.get_rank(k).name}"
+                    )
+                else:
+                    descr = "not provided"
+                return f"Legacy value: {descr}"
+
+            def native_val_info(k):
+                if k in native_values:
+                    val = native_values[k]
+                    descr = (
+                        f"{val} of type {type(val)}, from source {native_values.get_rank(k).name}"
+                    )
+                else:
+                    descr = "not provided"
+                return f"Native value: {descr}"
+
+            for key in sorted(legacy_values.get_keys() | native_values.get_keys()):
```

Comment:
Does this include passthrough args?

I'm not sure that question makes sense in this location... but I'm trying to work out if we're also validating the interpretation/consumption/tracking of the `pants ... -- these args here`.

Reply:
It does include passthru args. The way those works is, a string-list option can be registered as `passthrough=True`, and then the passthrough args are set as its value. This happens in parser.py and in native_options.py.